### DISCUSSION TITLE
Question about ServiceStack Auth on .NET Maui Blazor Hybrid

### DISCUSSION THREAD

**Discussion by Ryan-Britton** [ServiceStack/Discuss#105] [Q&A] [answered]
When implementing a .NET Maui Blazor Hybrid application intended to run on IOS/Android/Windows (reusing razor libraries for a web solution as well), I would love to use the great work that ServiceStack has done over the years to build up browser-based authentication flows for Apple, Google and Microsoft (using OpenID Connect/oAuth 2). A recent Blazor Tailwind demo showed a plethora of UI controls around authentication-centered functionality - but I see that uses the ASP.NET Identity Platform - and it seems that this is the recommended approach for ServiceStack apps going forward...is that the case? It adds additional complexity to the solution, since I am using OrmLite (rather then Entity Framework - and it seems a little messy to have 2 ORMs and 2 intertwined migration stories in a single app). What are your thoughts on all of that?

Also: the Blazor Web App Demo shows a solution which leverages the unified blazor hosting model for .NET 8. Would it be easy to port those wonderful identity controls over to a .NET MAUI Blazor Hybrid solution (where the server-part/API of the solution would be in a different project) so that I could use a browser-based authentication flow on startup of my app on Android/iOS/Windows?

**Comment by mythz** [comment, marked as answer]
>  and it seems that this is the recommended approach for ServiceStack apps going forward...is that the case?

That is the case, the easiest way to get a supported Auth Story that works in all of Blazor's Render Modes was to adopt ASP.NET Core Identity Auth and reuse all their existing Auth Integrations in the default Blazor templates.

> It adds additional complexity to the solution, since I am using OrmLite (rather then Entity Framework - and it seems a little messy to have 2 ORMs and 2 intertwined migration stories in a single app). What are your thoughts on all of that?

Using Entity Framework for the Identity User Models is effectively mandated in order to make use of all the existing functionality around ASP.NET Core Identity Auth and Identity UI. We also use OrmLite for the rest of the Application, and whilst there are 2 migrations solutions, OrmLite's migrations are more flexible which can perform schema and data migration updates that's also able to run Entity Framework's migrations so you still have a single command to setup a database after creating a schema and running both EF and OrmLite migrations, e.g:

```
npm run migrate
```

We no longer have full control over Auth which limits what features we can add, but at least all .NET Apps now use the same Auth Provider model to secure their Apps which stands a better chance for someone having documented a similar Auth workflow to what you want to do.

>  Would it be easy to port those wonderful identity controls over to a .NET MAUI Blazor Hybrid solution

I've honestly not looked at Maui Hybrid, which I thought was an Application around a Web View control in which case I would have thought all Blazor Components would be able to work as-is. But I don't really see MAUI gaining much momentum so I don't want to spend any time supporting it.

**Reply by Ryan-Britton** [reply]
Thanks for the guidance, sir. Could you elaborate on the "MAUI not gaining much momentum part"? What is the achilles heel here, in your opinion? What competitive approach do you feel is the best way to implement cross-platform apps with integrated mobile platform compatibility? I see quite a few Flutter examples in the ServiceStack docs - would that be your weapon of choice? Or do you think that maintaining separate codebases is worth the overhead?

**Reply by mythz** [reply]
IMO Microsoft has a poor history with .NET GUIs, which seem to be constantly abandoned and unmaintained, from WinForms, WPF, UWP, Silverlight, XNA, Windows Mobile, Xamarin, etc. I personally wouldn't invest in any of their GUIs until they use it to build some of the UIs for flagship products so they're forced not to abandoned it.

Yeah I'd personally use something else with a better ecosystem behind it like Flutter, React Native or Kotlin Multiplatform﻿ when it matures, or a PWA if it's suitable for your App.

**Reply by Ryan-Britton** [reply]
that's great advice, thanks.
